framework/include/linesearches/LineSearch.h

```diff
@@ -0,0 +1,47 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#ifndef LINESEARCH_H
+#define LINESEARCH_H
+
+#include "libmesh/parallel_object.h"
+#include "ConsoleStreamInterface.h"
+
+using namespace libMesh;
+
+class FEProblemBase;
+
+class LineSearch : public ConsoleStreamInterface, public ParallelObject
+{
+public:
+  LineSearch(FEProblemBase & fe_problem, MooseApp & app);
+
+  // A dummy pure virtual destructor because this must be a base class; the derived class must
+  // implement the line searching capabilities
+  virtual ~LineSearch() = 0;
+
+  /**
+   * writeable reference to number of non-linear iterations
+   */
+  size_t & nl_its() { return _nl_its; }
```

Comment:
Hmm... Why are you returning a writable reference here? What external class needs to change the internal state of this one? I see that you are changing it in the mechanical constraint but I don't understand why. It's always set to zero, but why not just initialize this member to zero? Maybe I'm missing something.

Reply:
Something has to zero this at the start of every non-linear solve

Reply:
Not sure where my email comment went. Instead, you should just have a method that zeros this out instead of exposing the inner member directly.